compendium/DeclarativeServices/src/manager/BundleLoader.cpp

```diff
@@ -72,62 +75,49 @@ GetComponentCreatorDeletors(const std::string& compName,
   if (bundleBinaries.lock()->count(bundleLoc) != 0u) {
     handle = bundleBinaries.lock()->at(bundleLoc);
   } else {
-#if defined(_WIN32)
-    std::wstring bundlePathWstr = UTF8StrToWStr(fromBundle.GetLocation());
-    handle = reinterpret_cast<void*>(LoadLibraryW(bundlePathWstr.c_str()));
-    if (handle == nullptr) {
-      std::error_code err_code(GetLastError(), std::generic_category());
-      std::string errMsg("Unable to load bundle binary ");
-      errMsg += fromBundle.GetLocation();
-      errMsg += ". Error: ";
-      errMsg += std::to_string(GetLastError());
-      throw cppmicroservices::SharedLibraryException(
-        err_code, std::move(errMsg), std::move(fromBundle));
-    }
+    SharedLibrary sh(bundleLoc);
+    try {
+#ifdef US_PLATFORM_POSIX
+      auto ctx = fromBundle.GetBundleContext();
+      auto opts = ctx.GetProperty(Constants::LIBRARY_LOAD_OPTIONS);
+      if (!opts.Empty()) {
+        sh.Load(any_cast<int>(opts));
+      } else {
+        sh.Load();
+      }
 #else
-    handle = dlopen(fromBundle.GetLocation().c_str(), RTLD_LAZY | RTLD_LOCAL);
-    if (handle == nullptr) {
-      std::error_code err_code(errno, std::generic_category());
-      std::string errMsg("Unable to load bundle binary ");
-      errMsg += fromBundle.GetLocation();
-      errMsg += ". Error: ";
-      const char* dlErrMsg = dlerror();
-      errMsg += (dlErrMsg) ? dlErrMsg : "none";
-      throw cppmicroservices::SharedLibraryException(
-        err_code, std::move(errMsg), std::move(fromBundle));
-    }
+      sh.Load();
 #endif
+    } catch (const std::system_error& ex) {
+      // SharedLibrary::Load() will throw a std::system_error when a shared library
+      // fails to load. Creating a SharedLibraryException here to throw with fromBundle information.
+      throw cppmicroservices::SharedLibraryException(ex.code(), ex.what(), std::move(fromBundle));
+    }
+    handle = sh.GetHandle();
     bundleBinaries.lock()->emplace(bundleLoc, handle);
   }
 
   const std::string symbolName =
     std::regex_replace(compName, std::regex("::"), "_");
   const std::string newInstanceFuncName("NewInstance_" + symbolName);
   const std::string deleteInstanceFuncName("DeleteInstance_" + symbolName);
-#if defined(_WIN32)
-  void* sym = reinterpret_cast<void*>(GetProcAddress(
-    reinterpret_cast<HMODULE>(handle), newInstanceFuncName.c_str()));
-  void* delsym = reinterpret_cast<void*>(GetProcAddress(
-    reinterpret_cast<HMODULE>(handle), deleteInstanceFuncName.c_str()));
+  
+  void* sym = fromBundle.GetSymbol(handle, newInstanceFuncName);
```

Comment:
Consider using a more descriptive file name (e.g., newsym or newInstSym)

Reply:
changed to newsym